internal/controller/metrics/metrics.go

```diff
@@ -1,42 +1,72 @@
 	"github.com/prometheus/client_golang/prometheus"
 	ctrlMetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
 )
 
 const (
-	shootName = "shootName"
-	state     = "state"
+	runtimeIDKeyName = "runtimeId"
+	state            = "state"
+	reason           = "reason"
+	runtimeIDLabel   = "kyma-project.io/runtime-id"
+	componentName    = "infrastructure_manager"
 )
 
-var (
-
-	//nolint:godox //TODO: test custom metric, remove when done with https://github.com/kyma-project/infrastructure-manager/issues/11
-	playgroundTotalReconciliationLoopsStarted = prometheus.NewCounter( //nolint:gochecknoglobals
-		prometheus.CounterOpts{
-			Name: "im_playground_reconciliation_loops_started_total",
-			Help: "Number of times reconciliation loop was started",
-		},
-	)
-
-	metricGardenerClustersState = prometheus.NewGaugeVec( //nolint:gochecknoglobals
-		prometheus.GaugeOpts{ //nolint:gochecknoglobals
-			Subsystem: "infrastructure_manager",
-			Name:      "im_gardener_clusters_state",
-			Help:      "Indicates the Status.state for GardenerCluster CRs",
-		}, []string{shootName, state})
-)
+type Metrics struct {
+	gardenerClustersStateGaugeVec *prometheus.GaugeVec
+}
 
-func init() {
-	ctrlMetrics.Registry.MustRegister(playgroundTotalReconciliationLoopsStarted, metricGardenerClustersState)
+func NewMetrics() Metrics {
+	m := Metrics{
+		gardenerClustersStateGaugeVec: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Subsystem: componentName,
+				Name:      "im_gardener_clusters_state",
+				Help:      "Indicates the Status.state for GardenerCluster CRs",
+			}, []string{runtimeIDKeyName, state, reason}),
+	}
+	ctrlMetrics.Registry.MustRegister(m.gardenerClustersStateGaugeVec)
+	return m
 }
 
-func IncrementReconciliationLoopsStarted() {
-	playgroundTotalReconciliationLoopsStarted.Inc()
+func (m Metrics) SetGardenerClusterStates(cluster v1.GardenerCluster) {
+	var runtimeID = cluster.GetLabels()[runtimeIDLabel]
+
+	if runtimeID != "" {
+		if len(cluster.Status.Conditions) != 0 {
+			var reason = cluster.Status.Conditions[0].Reason
+
+			// first clean the old metric
+			m.cleanUpGardenerClusterGauge(runtimeID)
+			m.gardenerClustersStateGaugeVec.WithLabelValues(runtimeID, string(cluster.Status.State), reason).Set(1)
+		}
+	}
 }
 
-func SetGardenerClusterStates(cluster v1.GardenerCluster) {
-	metricGardenerClustersState.WithLabelValues(cluster.Spec.Shoot.Name, string(cluster.Status.State)).Set(1)
+func (m Metrics) UnSetGardenerClusterStates(runtimeID string) {
+	m.cleanUpGardenerClusterGauge(runtimeID)
+}
+
+func (m Metrics) cleanUpGardenerClusterGauge(runtimeID string) {
+	var readyMetric, _ = m.gardenerClustersStateGaugeVec.GetMetricWithLabelValues(runtimeID, "Ready")
+	if readyMetric != nil {
+		readyMetric.Set(0)
+	}
+	var errorMetric, _ = m.gardenerClustersStateGaugeVec.GetMetricWithLabelValues(runtimeID, "Error")
+	if errorMetric != nil {
+		errorMetric.Set(0)
+	}
+	fmt.Printf("GardenerClusterStates set value to 0 for %v", runtimeID)
```

Comment:
Could we use logger for logging? Rest of the code uses `logr.Logger`

Reply:
I've removed this debug line